Improve WP Admin UI

westonruter opened the pull request.

westonruter committed:
Add author post type support

westonruter committed:
Remove Add New post link for snapshots; harden capabilities

westonruter committed:
Show only dirty settings in edit post screen; use UUID as metabox title

westonruter committed:
Restore UUID as heading so it is selectable

valendesigns commented on php/class-customize-snapshot-manager.php:
```diff
@@ -284,7 +289,7 @@ public function create_post_type() {
 	function setup_metaboxes() {
 		$id = self::POST_TYPE;
-		$title = __( 'Data', 'customize-snapshots' );
+		$title = __( 'Data', 'customize-snapshot' );
```
The text domain is `customize-snapshots`

westonruter committed:
Fix text domain

valendesigns commented on php/class-plugin.php:
```diff
@@ -74,13 +74,23 @@ public function register_styles( \WP_Styles $wp_styles ) {
 			$allcaps['customize_publish'] = true;
 		}
 
+		// Grant all customize snapshot caps which weren't explicitly disallowed to users who can customize.
```
This is causing test failures.

westonruter replied:
I restored the initial cap check in ab6ca42. Tests are passing now.

valendesigns commented on php/class-customize-snapshot-manager.php:
```diff
@@ -494,9 +502,6 @@ public function update_snapshot() {
 		if ( ! check_ajax_referer( self::AJAX_ACTION, 'nonce', false ) ) {
 			status_header( 400 );
 			wp_send_json_error( 'bad_nonce' );
-		} elseif ( ! current_user_can( 'customize' ) ) {
```
Actually this is what is causing the test failures. Not sure why you want to remove this?

westonruter replied:
@valendesigns because I added it below instead. The `create_posts` cap is now mapped to `do_not_allow` in the post type so as to hide the “Add New” links for snapshots in the admin. So then we have to use `customize` cap for this instead directly here, which would be doing below anyway.

valendesigns replied:
You will never get pass this condition `} elseif ( empty( $this->unsanitized_snapshot_post_data ) ) {` because `current_user_can( 'customize' )` will be false in `Customize_Snapshot_Manager::capture_unsanitized_snapshot_post_data`

westonruter replied:
I restored the `customize` cap check.

valendesigns committed:
Allow non-dirty settings to be revealed

valendesigns committed:
Check that array index isset

westonruter committed:
Restore initial customize check

westonruter committed:
Restore edit_published_posts and delete_published_posts caps for customize snapshots

We'll need a different mechanism to prevent mutating published (non-future) snapshots

westonruter committed:
Add quick edit link to open Customizer

westonruter committed:
Add Open in Customizer button to edit snapshot screen

valendesigns commented on php/class-customize-snapshot-manager.php:
```diff
@@ -348,6 +348,16 @@ function render_data_metabox( $post ) {
+			'<p><a href="%s" class="button button-secondary">%s</a></p>',
+			esc_url( $customize_url ),
+			esc_html__( 'Open in Customizer', 'customize-snapshots' )
+		);
```
This commit makes me happy 😄

westonruter committed:
Hide Snapshot Customize links if status is publish

valendesigns commented:
@westonruter When you publish the snapshot all the settings disappear from the metabox

valendesigns commented:
Oh they are hidden after published. So we need to remove the checkbox toggle and hidden attribute for published snapshots so all the settings display

westonruter commented:
@valendesigns I think the problem is that all of the settings get sent with `dirty: false` at the moment of Publish. We should preserve the original `dirty` state.

westonruter committed:
Add missing js-doc return tags

westonruter committed:
Set isPreview once snapshot saved; send original _dirty state

westonruter committed:
Make sure Published button can be re-disabled when publishing immediately after saving a snapsot

westonruter committed:
Move Open in Customizer button to misc publish box

westonruter committed:
Move Open in Customizer button to post_submitbox_minor_actions

westonruter committed:
Remove publish metabox got published snapshots since they should be immutable

westonruter committed:
Hide trash and quick edit links with view instead of edit for published snapshots

westonruter committed:
Disable author metabox for published snapshots

westonruter committed:
Merge branch 'bugfix/replace-state' of https://github.com/xwp/wp-customize-snapshots into feature/wp-admin

westonruter committed:
Merge branch 'release/0.4.0' of https://github.com/xwp/wp-customize-snapshots into feature/wp-admin

westonruter committed:
Update wp-dev-lib 159119b...0f3058b: Merge pull request xwp/wp-dev-lib#187 from branch bugfix/nested-phpunit-support2

https://github.com/xwp/wp-dev-lib/compare/159119b...0f3058b

westonruter commented:
@valendesigns ok, ready for you to review.

valendesigns commented:
@westonruter I think we should hide the "Save Draft" button or change the customize button to "Edit" so it fits in the space. Thoughts?

valendesigns commented:
We also need to handle the publish button so the Snapshot updates the front-end. Otherwise scheduling will not work.

valendesigns commented:
@westonruter An issue with the updated state replacement JS is that the `?` is being left behind when there are no params in the url. It should check that there are params and remove the question mark if there aren't any.

westonruter committed:
Trim trailing question mark

westonruter commented:
@valendesigns 

> I think we should hide the "Save Draft" button or change the customize button to "Edit" so it fits in the space. Thoughts? We also need to handle the publish button so the Snapshot updates the front-end. Otherwise scheduling will not work.

Actually, perhaps we should just hide the publish box entirely. This would be what #15 would implement, and at that time, the publish button could be restored.

westonruter commented:
I think I know what to do. We can just remove the `publish_posts` cap for now and it will just work 😄

westonruter committed:
Remove publish_posts cap until #15 is implmented

westonruter commented:
@valendesigns oh, and you know what? Now with b3ea60a it would be sweet if when a user lacks `customize_publish` capability, in addition to a Save button there could be a “Submit” button that serves the same purpose as the “Submit for Review” button when a user wants to submit a post as pending status:

![image](https://cloud.githubusercontent.com/assets/134745/15950613/66c40ff4-2e65-11e6-8b76-97e996457adb.png)

valendesigns commented:
So it makes the publish button say submit for review?

westonruter committed:
Introduce a submit button if user cannot publish

westonruter commented:
@valendesigns yeah, try this: https://github.com/xwp/wp-customize-snapshots/compare/feature/wp-admin...feature/submit-for-review

westonruter committed:
Prevent post_name from being striped when submitting for review

westonruter committed:
Move Open in Customizer back to metabox until publish metabox is supported

westonruter committed:
Hide Customize quicklink if user cannot edit post

westonruter committed:
Hide edit bulk action, quick edit, and author metabox

westonruter committed:
Restore delete quick edit link

westonruter committed:
Add post info from submitbox to data metabox

westonruter committed:
Restrict editing and deleting others' posts

westonruter committed:
Merge branch 'feature/submit-for-review' into feature/wp-admin

westonruter commented:
@valendesigns I merged `feature/submit-for-review` into this branch. I'm feeling really good about it now. I think I addressed all of your concerns.

westonruter committed:
Add test for preserving post name when submitting as pending

valendesigns committed:
Fix the perms message and add new changeButton method to reduce duplication
